File: schedule_structures.py
from dataclasses import dataclass, field
from typing import List

@dataclass
class Room:
    id: str
    type: str

@dataclass
class Teacher:
    name: str

@dataclass
class SchoolClass:
    name: str

@dataclass
class Activity:
    teacher: Teacher
    school_class: SchoolClass
    subject: str
    weekly_count: int

@dataclass
class Lesson:
    day: int
    slot: int
    room: Room
    activity: Activity

@dataclass
class Timetable:
    rooms: List[Room] = field(default_factory=list)
    activities: List[Activity] = field(default_factory=list)
    lessons: List[Lesson] = field(default_factory=list)

    def add_room(self, room: Room):
        self.rooms.append(room)

    def add_activity(self, activity: Activity):
        self.activities.append(activity)

    def add_lesson(self, lesson: Lesson):
        self.lessons.append(lesson)

File: schedule_structures.R
# Basic R6 data structures for a timetable generator

library(R6)

Room <- R6Class("Room",
  public = list(
    id = NULL,
    type = NULL,
    initialize = function(id, type) {
      self$id <- id
      self$type <- type
    }
  )
)

Teacher <- R6Class("Teacher",
  public = list(
    name = NULL,
    initialize = function(name) {
      self$name <- name
    }
  )
)

SchoolClass <- R6Class("SchoolClass",
  public = list(
    name = NULL,
    initialize = function(name) {
      self$name <- name
    }
  )
)

Activity <- R6Class("Activity",
  public = list(
    teacher = NULL,
    school_class = NULL,
    subject = NULL,
    weekly_count = NULL,
    initialize = function(teacher, school_class, subject, weekly_count) {
      self$teacher <- teacher
      self$school_class <- school_class
      self$subject <- subject
      self$weekly_count <- weekly_count
    }
  )
)

Lesson <- R6Class("Lesson",
  public = list(
    day = NULL,
    slot = NULL,
    room = NULL,
    activity = NULL,
    initialize = function(day, slot, room, activity) {
      self$day <- day
      self$slot <- slot
      self$room <- room
      self$activity <- activity
    }
  )
)

Timetable <- R6Class("Timetable",
  public = list(
    rooms = list(),
    activities = list(),
    lessons = list(),
    initialize = function() {
      self$rooms <- list()
      self$activities <- list()
      self$lessons <- list()
    },
    add_room = function(room) {
      self$rooms[[length(self$rooms) + 1]] <- room
    },
    add_activity = function(activity) {
      self$activities[[length(self$activities) + 1]] <- activity
    },
    add_lesson = function(lesson) {
      self$lessons[[length(self$lessons) + 1]] <- lesson
    }
  )
)

